scipy/cluster/_optimal_leaf_ordering.pyx

```diff
@@ -0,0 +1,495 @@
+
+                        # For that chosen (u, w), record
+                        #    the resulting minimal M[u, w] = M[u, m] + M[k, w] + D[m, k]
+                        M[u, w] = cur_min_M
+                        M[w, u] = cur_min_M
+                        #    wether we need to swap V_l and V_r given the current chosen (m, k)
+                        #    (see note above)
+                        #    this saves us from storing (m, k) and doing back-tracking later. 
+                        swap_status[u, w, 0] = swap_L
+                        swap_status[w, u, 0] = swap_L
+                        swap_status[u, w, 1] = swap_R
+                        swap_status[w, u, 1] = swap_R
+
+                        # We are getting a fresh `w` so need to resort M[{k}, w]
+                        free(k_vals)
+                        free(k_idx)
+
+                    # We are getting a fresh `u` so need to resort M[{m}, u]
+                    free(m_vals)
+                    free(m_idx)
+
+        # We are about to get fresh clusters.
+        free(u_clusters)
+        free(m_clusters)
+        free(w_clusters)
+        free(k_clusters)
+
+        # We are now ready to find the best minimal value for M[{u}, {w}]
+        cur_min_M = 1073741824.0 #2^30
+        for u in range(v_l_min, v_l_max):
+            for w in range(v_r_min, v_r_max):
+                if M[u, w] < cur_min_M:
+                    cur_min_M = M[u, w]
+                    best_u = u
+                    best_w = w
+
+        # If v_l, v_r are not singletons, record wether our choice of (u, w) for V requires a swap of it's children.
+        if v_l >= n_points:
+            must_swap[v_l - n_points] = int(swap_status[best_u, best_w, 0])
+        if v_r >= n_points:
+            must_swap[v_r - n_points] = int(swap_status[best_u, best_w, 1])
+
+    return must_swap
+
+
+def optimal_leaf_ordering(Z, D):
+    """
+    Compute the optimal leaf order for Z (according to D)
+    and return an optimally sorted Z. 
+
+    We start by sorting and relabelling Z and D according to the
+    current leaf order in Z.
+    
+    This is because when everything is sorted each cluster (including singletons)
+    can be defined by its range over (0...n_points).
+
+    This is used extensively loop efficiently over the various arrays in the algorithm.
+
+    """
+    if not is_valid_linkage(Z):
+        raise("Not a valid linkage map. ")
+    
+    if is_valid_y(D):
+        sorted_D = squareform(D)
+    elif is_valid_dm(D):
+        sorted_D = D
+        pass
+    else:
+        raise("Not a valid distance matrix (neither condensed nor square forms)")
+
+    n_points = Z.shape[0] + 1
+    n_clusters = 2*n_points - 1
+
+    # Get the current linear ordering
+    sorted_leaves = leaves_list(Z)
+
+    # Create map from original order to sorted order.
+    original_order_to_sorted_order = dict((orig_i, sorted_i) for sorted_i, orig_i in enumerate(sorted_leaves))
```

Comment:
style: lines should be <80 chars (pep8), could you change that in this whole file?